### PR TITLE
Added rclcpp component to Republish

### PR DESCRIPTION
**Pull request by ahcorde**
These changes allow to use the republisher as a `rclcpp_component` and as a normal node.

An example of how to use this:

```bash
ros2 run image_transport republish --in_transport compressed in/compressed:=/camera/image/compressed out/compressed:=/camera/republish/compressed out:=/camera/republish/image_raw
```

### CODE REVIEW

File: image_transport/include/image_transport/republish.hpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include "image_transport/visibility_control.hpp"
+
+#include <rclcpp/rclcpp.hpp>
+
+namespace image_transport
+{
+class Republisher : public rclcpp::Node
+{
+public:
+  //! Constructor
```
Comment (adityapande-1995) [inline image_transport/include/image_transport/republish.hpp]: Is the "!" a typo ?

**Review by adityapande-1995 — COMMENTED** [review]
Looks good ! Could you add a test or an example ? Maybe here : https://github.com/ros-perception/image_common/tree/noetic-devel/image_transport/tutorial/src ?

File: image_transport/include/image_transport/republish.hpp (hunk 1 above)
Comment (ahcorde) [reply image_transport/include/image_transport/republish.hpp]: https://github.com/ros-perception/image_common/pull/275/commits/68cb1f50c9102163946e233e1e6d8334d0aa5712

**Review by adityapande-1995 — APPROVED** [review]
Looks good ! Is there a way to add a test here ?

File: image_transport/src/republish.cpp [hunk 2]
```diff
@@ -71,8 +108,8 @@ int main(int argc, char ** argv)
     });
 
     pub_options.qos_overriding_options = qos_override_options;
-    auto pub = image_transport::create_publisher(
-      node.get(), out_topic,
+    this->pub = image_transport::create_publisher(
+      this->shared_from_this().get(), out_topic,
```
Comment (mikeferguson) [inline image_transport/src/republish.cpp]:
```suggestion
      this, out_topic,
```

File: image_transport/src/republish.cpp [hunk 3]
```diff
@@ -83,31 +120,34 @@ int main(int argc, char ** argv)
     rclcpp::SubscriptionOptions sub_options;
     sub_options.qos_overriding_options = qos_override_options;
 
-    auto sub = image_transport::create_subscription(
-      node.get(), in_topic, std::bind(pub_mem_fn, &pub, std::placeholders::_1),
+    this->sub = image_transport::create_subscription(
+      this->shared_from_this().get(), in_topic, std::bind(pub_mem_fn, &pub, std::placeholders::_1),
```
Comment (mikeferguson) [inline image_transport/src/republish.cpp]:
```suggestion
      this, in_topic, std::bind(pub_mem_fn, &pub, std::placeholders::_1),
```

File: image_transport/src/republish.cpp [hunk 4]
```diff
@@ -83,31 +120,34 @@ int main(int argc, char ** argv)
...
-
     // Load transport plugin
     typedef image_transport::PublisherPlugin Plugin;
     pluginlib::ClassLoader<Plugin> loader("image_transport", "image_transport::PublisherPlugin");
     std::string lookup_name = Plugin::getLookupName(out_transport);
 
     auto instance = loader.createUniqueInstance(lookup_name);
     std::shared_ptr<Plugin> pub = std::move(instance);
-    pub->advertise(node.get(), out_topic);
+    pub->advertise(this->shared_from_this().get(), out_topic);
```
Comment (mikeferguson) [inline image_transport/src/republish.cpp]:
```suggestion
    pub->advertise(this, out_topic);
```

File: image_transport/src/republish.cpp [hunk 5]
```diff
@@ -83,31 +120,34 @@ int main(int argc, char ** argv)
...
 
     // Use PublisherPlugin::publish as the subscriber callback
     typedef void (Plugin::* PublishMemFn)(const sensor_msgs::msg::Image::ConstSharedPtr &) const;
     PublishMemFn pub_mem_fn = &Plugin::publishPtr;
-    auto sub = image_transport::create_subscription(
-      node.get(), in_topic,
+    this->sub = image_transport::create_subscription(
+      this->shared_from_this().get(), in_topic,
```
Comment (mikeferguson) [inline image_transport/src/republish.cpp]:
```suggestion
      this, in_topic,
```

File: image_transport/src/republish_program.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    out_transport = image_transport::get_option(args, "--out_transport");
+  }
+
+  rclcpp::NodeOptions options;
+  // override default parameters with the desired transform
+  options.parameter_overrides(
+  {
+    {"in_transport", in_transport},
+    {"out_transport", out_transport},
+  });
```
Comment (mikeferguson) [inline image_transport/src/republish_program.cpp]: What is the point of all this? It seems to be the only specialized thing you are doing over the default behavior of generating an executable using rclcpp_components_register_node

File: image_transport/CMakeLists.txt [hunk 7]
```diff
@@ -58,11 +59,33 @@ target_link_libraries(list_transports
...
+)
+target_link_libraries(republish_node
+  ${PROJECT_NAME}
+)
+ament_target_dependencies(republish_node
+  pluginlib
+  rclcpp_components
+  rclcpp
+)
+rclcpp_components_register_nodes(republish_node "image_transport::Republisher")
```
Comment (mikeferguson) [inline image_transport/CMakeLists.txt]: You could actually remove the entire add_executable below, delete republish_program.cpp and the utilities files and generate a node that loads the plugin and runs it with:

```
rclcpp_components_register_node(republish_node
  PLUGIN "image_transport::Republisher"
  EXECUTABLE republish
)
```

**Review by mikeferguson — COMMENTED** [review]
I don't understand why you go through manually processing the parameters?

**Comment by ahcorde** [comment]
When using `--out_transport` the app is crashing.

File: image_transport/src/republish.cpp [hunk 8]
```diff
@@ -83,31 +120,34 @@ int main(int argc, char ** argv)
...
-
     // Load transport plugin
     typedef image_transport::PublisherPlugin Plugin;
     pluginlib::ClassLoader<Plugin> loader("image_transport", "image_transport::PublisherPlugin");
     std::string lookup_name = Plugin::getLookupName(out_transport);
 
     auto instance = loader.createUniqueInstance(lookup_name);
     std::shared_ptr<Plugin> pub = std::move(instance);
-    pub->advertise(node.get(), out_topic);
+    pub->advertise(this, out_topic);
```
Comment (mikeferguson) [inline image_transport/src/republish.cpp]: should this be this->pub? rather than declaring a new one in the line above? Might that be your crash?

**Comment by ahcorde** [comment]
Yep, but I have now this awesome message

```bash
Warning: class_loader.ClassLoader: SEVERE WARNING!!! Attempting to unload library while objects created by this loader exist in the heap! You should delete your objects before attempting to unload the library or destroying the ClassLoader. The library will NOT be unloaded.
         at line 127 in /home/ahcorde/ros2_rolling/src/ros/class_loader/src/class_loader.cpp
```

**Comment by ahcorde** [comment]
Error is fixed too. The loader need to keep alive too

**Comment by mikeferguson** [comment]
Honestly - probably 70% of the ROS codebase that uses pluginlib has that error on shutdown

**Review by mikeferguson — APPROVED** [review]

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20301/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14712/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21036/)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20327/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14738/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21071/)

**Comment by asimonov** [comment]
what is interesting is that it broke existing 'republish' functionality in Jazzy.

for anyone who ended up here scratching their head why their republish does not work after upgrading to Jazzy, here is the answer:

this used to work (using launch xml syntax):

```
    <node pkg="image_transport"
          exec="republish"
          name="republish_compressed_left"
          args="compressed raw --ros-args --remap in/compressed:=/camera/left/image_raw/compressed --remap out:=/camera/left/image_raw"
          output="both">
    </node>
    <node pkg="image_transport"
          exec="republish"
          name="republish_compressed_right"
          args="compressed raw --ros-args --remap in/compressed:=/camera/right/image_raw/compressed --remap out:=/camera/right/image_raw"
          output="both">
    </node>
```

now with the move from node to component the following has to be used to provide exactly the same functionality.

```
    <node_container pkg="rclcpp_components" 
                    exec="component_container" 
                    name="republish_container" 
                    namespace="">

        <composable_node pkg="image_transport" 
                         plugin="image_transport::Republisher" 
                         name="republish_left" 
                         namespace="">
            <param name="in_transport"             value="compressed" />
            <param name="out_transport"            value="raw" />
            <remap from="in/compressed"            to="/camera/left/image_raw/compressed"/>
            <remap from="out"                      to="/camera/left/image_raw"/>
            <extra_arg name="use_intra_process_comms" value="true" />
        </composable_node>
        <composable_node pkg="image_transport" 
                         plugin="image_transport::Republisher" 
                         name="republish_right" 
                         namespace="">
            <param name="in_transport"             value="compressed" />
            <param name="out_transport"            value="raw" />
            <remap from="in/compressed"            to="/camera/right/image_raw/compressed"/>
            <remap from="out"                      to="/camera/right/image_raw"/>
            <extra_arg name="use_intra_process_comms" value="true" />
        </composable_node>

    </node_container>
```


and this is not documented anywhere, which is very very frustrating

the other bad aspect of how the change was made is the executable name was left 'republish', but one cannot use it as node anymore. I would say the existing node 'republish' should have been left alone, and 'republish_component' created in addition to existing functionality.

**Comment by mikeferguson** [comment]
I think you just need to change ```exec="republish"``` to ```exec="republish_node"``` and your original will work.

**Pull request merged**